pkg/db-repo/change_history_manager.go

```diff
@@ -0,0 +1,232 @@
+	}
+
+	for _, trigger := range triggers {
+		c.appendStatement(fmt.Sprintf(`DROP TRIGGER IF EXISTS %s`, trigger))
+	}
+}
+
+func (c *changeHistoryManager) createHistoryTriggers() {
+	const NewRecord = "NEW"
+	const OldRecord = "OLD"
+
+	c.appendStatement(fmt.Sprintf(`CREATE TRIGGER %s_ai AFTER INSERT ON %s FOR EACH ROW %s WHERE %s`,
+		c.originalTable.tableName,
+		c.originalTable.tableName,
+		c.insertHistoryEntry("insert", true),
+		c.primaryKeysMatchCondition(NewRecord),
+	))
+
+	c.appendStatement(fmt.Sprintf(`CREATE TRIGGER %s_au AFTER UPDATE ON %s FOR EACH ROW %s WHERE %s AND (%s)`,
+		c.originalTable.tableName,
+		c.originalTable.tableName,
+		c.insertHistoryEntry("update", true),
+		c.primaryKeysMatchCondition(NewRecord),
+		c.rowUpdatedCondition(),
+	))
+
+	c.appendStatement(fmt.Sprintf(`CREATE TRIGGER %s_bd BEFORE DELETE ON %s FOR EACH ROW %s WHERE %s`,
+		c.originalTable.tableName,
+		c.originalTable.tableName,
+		c.insertHistoryEntry("delete", false),
+		c.primaryKeysMatchCondition(OldRecord),
+	))
+
+	c.appendStatement(fmt.Sprintf(`CREATE TRIGGER %s_revai BEFORE INSERT ON %s FOR EACH ROW %s`,
+		c.historyTable.tableName,
+		c.historyTable.tableName,
+		c.incrementRevision(),
+	))
+}
+
+func (c *changeHistoryManager) insertHistoryEntry(action string, includeAuthorEmail bool) string {
+	columnNames := c.originalTable.columnNamesQuoted()
+	if !includeAuthorEmail {
+		columnNames = c.originalTable.columnNamesQuotedExcludingValue(c.settings.ChangeAuthorField)
+	}
+
+	columns := strings.Join(columnNames, ",")
+	values := "d." + strings.Join(columnNames, ", d.")
+
+	return fmt.Sprintf(`
+		INSERT INTO %s (change_history_action,change_history_revision,change_history_action_at,%s) 
+			SELECT '%s', NULL, NOW(), %s 
+			FROM %s AS d`,
+		c.historyTable.tableNameQuoted,
+		columns,
+		action,
+		values,
+		c.originalTable.tableNameQuoted)
+}
+
+func (c *changeHistoryManager) incrementRevision() string {
+	return fmt.Sprintf(`
+		BEGIN 
+			SET NEW.change_history_revision = (SELECT IFNULL(MAX(d.change_history_revision), 0) + 1 FROM %s as d WHERE %s); 
+		END`,
+		c.historyTable.tableNameQuoted,
+		c.primaryKeysMatchCondition("NEW"),
+	)
+}
+
+func (c *changeHistoryManager) primaryKeysMatchCondition(record string) string {
+	var conditions []string
+	for _, columnName := range c.originalTable.primaryKeyNamesQuoted() {
+		condition := fmt.Sprintf("d.%s = %s.%s", columnName, record, columnName)
+		conditions = append(conditions, condition)
+	}
+	return strings.Join(conditions, " AND ")
+}
+
+func (c *changeHistoryManager) rowUpdatedCondition() string {
+	columnNames := c.originalTable.columnNamesQuotedExcludingValue(c.settings.ChangeAuthorField)
+	var conditions []string
+	for _, columnName := range columnNames {
+		condition := fmt.Sprintf("OLD.%s != NEW.%s", columnName, columnName)
+		conditions = append(conditions, condition)
+	}
+	return strings.Join(conditions, " OR ")
+}
+
+func (c *changeHistoryManager) updateHistoryTable() bool {
```

Comment:
not sure how this will be in production later. we dont do auto migrations there...

Reply:
this auto migration is safe because it only adds columns.
But you're right, maybe we need a way to disable this from the configuration.
Let's wait for @j4k4 opinion about this.